### PR TITLE
refactor(fine-tuning): enhance error handling and documentation

### LINKED ISSUE #14: Hey @leonvanbokhorst - I've reviewed your changes - here's some feedback:
Hey @leonvanbokhorst - I've reviewed your changes - here's some feedback:

**Overall Comments**:
- Consider adding error handling for dataset loading and model initialization failures, particularly around MPS device availability
- Please add documentation explaining the rationale behind the specific training parameter values chosen for M3 optimization (batch size, gradient accumulation, learning rate etc)
- The main execution flow in inference_example could be restructured more clearly - consider separating the model loading and inference logic

<details>
<summary>Here's what I looked at during the review</summary>

- 🟢 **General issues**: all looks good
- 🟢 **Security**: all looks good
- 🟢 **Testing**: all looks good
- 🟡 **Complexity**: 1 issue found
- 🟢 **Documentation**: all looks good
</details>

***

<details>
<summary>Sourcery is free for open source - if you like our reviews please consider sharing them ✨</summary>

- [X](https://twitter.com/intent/tweet?text=I%20just%20got%20an%20instant%20code%20review%20from%20%40SourceryAI%2C%20and%20it%20was%20brilliant%21%20It%27s%20free%20for%20open%20source%20and%20has%20a%20free%20trial%20for%20private%20code.%20Check%20it%20out%20https%3A//sourcery.ai)
- [Mastodon](https://mastodon.social/share?text=I%20just%20got%20an%20instant%20code%20review%20from%20%40SourceryAI%2C%20and%20it%20was%20brilliant%21%20It%27s%20free%20for%20open%20source%20and%20has%20a%20free%20trial%20for%20private%20code.%20Check%20it%20out%20https%3A//sourcery.ai)
- [LinkedIn](https://www.linkedin.com/sharing/share-offsite/?url=https://sourcery.ai)
- [Facebook](https://www.facebook.com/sharer/sharer.php?u=https://sourcery.ai)

</details>

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

_Originally posted by @sourcery-ai[bot] in https://github.com/leonvanbokhorst/model-adaptation/pull/13#pullrequestreview-2404670940_

### PR DESCRIPTION
**Pull request by leonvanbokhorst**
- Add comprehensive error handling with try/except blocks
- Improve docstrings with return types, args and raises sections
- Optimize device selection logic for CUDA/MPS/CPU
- Update tokenizer configuration to use PyTorch tensors
- Add detailed training parameter documentation for M3 architecture
- Implement proper type hints for function signatures
Fixes #14

## Summary by Sourcery

Enhance error handling and documentation in the fine-tuning module. Improve device selection logic and implement type hints for better code clarity. Update docstrings and add detailed documentation for training parameters.

Enhancements:
- Enhance error handling by adding try/except blocks to critical functions.
- Optimize device selection logic for CUDA, MPS, and CPU to improve performance.
- Implement proper type hints for function signatures to improve code clarity.

Documentation:
- Improve docstrings by adding return types, arguments, and raises sections for better documentation.
- Add detailed documentation for training parameters specific to the M3 architecture.

### CODE REVIEW

File: src/04_fine-tuning.py [hunk 1]
```diff
@@ -46,16 +46,32 @@ def initialize_tokenizer(model_name: str, hf_token: str) -> AutoTokenizer:
...
-        model_name, token=HF_TOKEN, device_map="auto", torch_dtype=torch.float16
-    )
+    try:
+        # Determine optimal device
+        device = "cuda" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "cpu"
+        
+        model = AutoModelForCausalLM.from_pretrained(
+            model_name, 
+            token=HF_TOKEN,
+            device_map=device,
```
[the remark this answers is not in the document]
Comment (leonvanbokhorst) [reply src/04_fine-tuning.py:70]: Created an issue for this.
